common/range-map/src/range_wrapper.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::cmp::Ordering;
+use core::ops::Range;
+
+#[derive(Eq, Debug, Clone)]
+pub struct RangeWrapper<T> {
+    pub range: Range<T>,
+}
+
+impl<T> RangeWrapper<T> {
+    pub fn new(range: Range<T>) -> RangeWrapper<T> {
+        RangeWrapper { range }
+    }
+}
+
+impl<T> PartialEq for RangeWrapper<T>
+where T: Eq
+{
+    fn eq(&self, other: &RangeWrapper<T>) -> bool {
+        self.range.start == other.range.start && self.range.end == other.range.end
+    }
+}
+
+impl<T> Ord for RangeWrapper<T>
+where T: Ord + std::fmt::Debug
+{
+    fn cmp(&self, other: &RangeWrapper<T>) -> Ordering {
+        let start = self.range.start.cmp(&other.range.start);
+        let end = self.range.end.cmp(&other.range.end);
+
+        if other.range.start == other.range.end {
+            if end == Ordering::Less {
+                return Ordering::Less;
+            }
+
+            return Ordering::Greater;
+        }
+
+        match start {
+            Ordering::Less => match end {
+                Ordering::Less => {
+                    return Ordering::Less;
+                }
+                Ordering::Equal => {
+                    return Ordering::Less;
+                }
+                Ordering::Greater => {
+                    return Ordering::Greater;
+                }
+            },
+            Ordering::Equal => {
+                return end;
+            }
+            Ordering::Greater => match end {
+                Ordering::Less => {
+                    return Ordering::Less;
+                }
+                Ordering::Equal => {
+                    return Ordering::Greater;
+                }
+                Ordering::Greater => {
+                    return Ordering::Greater;
+                }
+            },
+        }
+    }
```

Comment:
These lines are just a complex way of comparing `(self.range.end, self.range.start)` and `(other.end, other.start)`.